soroban-spec/src/gen/json.rs

```diff
@@ -0,0 +1,310 @@
+}
+
+impl TryFrom<&ScSpecUdtStructFieldV0> for StructField {
+    type Error = Error;
+
+    fn try_from(f: &ScSpecUdtStructFieldV0) -> Result<Self, Self::Error> {
+        let name = f.name.to_string()?;
+        let value = SpecType::try_from(&f.type_)?;
+        Ok(StructField {
+            name: name,
+            value: value,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FunctionInput {
+    name: String,
+    value: SpecType,
+}
+
+impl TryFrom<&ScSpecFunctionInputV0> for FunctionInput {
+    type Error = Error;
+
+    fn try_from(f: &ScSpecFunctionInputV0) -> Result<Self, Self::Error> {
+        let name = f.name.to_string()?;
+        let value = SpecType::try_from(&f.type_)?;
+        Ok(FunctionInput {
+            name: name,
+            value: value,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct UnionCase {
+    name: String,
+    value: Option<SpecType>,
+}
+
+impl TryFrom<&ScSpecUdtUnionCaseV0> for UnionCase {
+    type Error = Error;
+
+    fn try_from(c: &ScSpecUdtUnionCaseV0) -> Result<Self, Self::Error> {
+        let name = c.name.to_string()?;
+        let value = c.type_.as_ref().map(SpecType::try_from);
+        Ok(UnionCase {
+            name: name,
+            value: match value {
+                None => None,
+                Some(r) => Some(r?),
+            },
+        })
+    }
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize)]
+#[serde(tag = "type")]
+#[serde(rename_all = "camelCase")]
+pub enum SpecType {
+    U64,
+    I64,
+    U32,
+    I32,
+    Bool,
+    Symbol,
+    Bitset,
+    Status,
+    Bytes,
+    BigInt,
+    Function {
+        name: String,
+        inputs: Vec<FunctionInput>,
+        outputs: Vec<SpecType>,
+    },
+    Map {
+        key: Box<SpecType>,
+        value: Box<SpecType>,
+    },
+    Option {
+        value: Box<SpecType>,
+    },
+    Result {
+        value: Box<SpecType>,
+        error: Box<SpecType>,
+    },
+    Set {
+        element: Box<SpecType>,
+    },
+    Vec {
+        element: Box<SpecType>,
+    },
+    Tuple {
+        elements: Vec<SpecType>,
+    },
+    UserDefined {
+        name: String,
+    },
+    Struct {
+        name: String,
+        fields: Vec<StructField>,
+    },
+    Union {
+        name: String,
+        cases: Vec<UnionCase>,
+    },
+}
+
+impl TryFrom<&ScSpecTypeDef> for SpecType {
+    type Error = Error;
+
+    fn try_from(spec: &ScSpecTypeDef) -> Result<Self, Self::Error> {
+        match spec {
+            ScSpecTypeDef::Map(inner) => {
+                let key = SpecType::try_from(inner.key_type.as_ref());
+                let value = SpecType::try_from(inner.value_type.as_ref());
+                Ok(SpecType::Map {
+                    key: Box::new(key?),
+                    value: Box::new(value?),
```

Comment:
I'm going to make a quick change to this PR before merging. It is helpful if the `?`'s appear near the thing that can fail, so rather than putting `?` on `r` it's helpful if it exists up near the try_from.